src/graph.cpp

```diff
@@ -165,9 +165,66 @@ bool Graph::path(const std::list<std::string> &source, const std::string &target
     return false;
 }
 
+std::list<Edge *> Graph::bellmanFord(const std::string &source, const std::string &target) {
+    for (Vertex *v: vertexSet) {
+        v->setCost(UINT32_MAX);
+        v->setPath(nullptr);
+    }
+    findVertex(source)->setCost(0);
+
+    for (int i = 1; i <= vertexSet.size(); i++) { //V times
+        for (Vertex *v: vertexSet) { //Relax every Edge
+            for (Edge *e: v->getIncoming()) {
+                if (e->getCapacity() > 0) {
+                    int tempCost = e->getOrig()->getCost() + e->getCost();
+                    if (tempCost < v->getCost()) {
+                        if (i == vertexSet.size()) { //Edge being relaxed on Nth iteration - Negative cycle!
+                            Vertex *currentVertex = v;
+                            std::list<Edge *> negativeCycle;
+                            do {
+                                negativeCycle.push_back(currentVertex->getPath());
+                                currentVertex = currentVertex->getPath()->getOrig();
+                            } while (currentVertex != v);
+                            return negativeCycle;
+                        }
+                        v->setCost(tempCost);
+                        v->setPath(e);
+                    }
+                }
+            }
+        }
+    }
+    return {};
+}
+
+std::pair<unsigned int, unsigned int>
+Graph::minCostMaxFlow(const std::string &source, const std::string &target, Graph &residualGraph) {
+    std::pair<unsigned int, unsigned int> result;
+    result.first = edmondsKarp({source}, target, residualGraph);
+
+    Graph minCostResidual;
+    makeMinCostResidual(minCostResidual);
+
+    std::list<Edge *> negativeCycle = minCostResidual.bellmanFord(source, target);
+    while (!negativeCycle.empty()) {
+        unsigned int bottleneckCapacity = minCostResidual.findListBottleneck(negativeCycle);
+        minCostResidual.augmentMinCostPath(negativeCycle, bottleneckCapacity);
+        negativeCycle = bellmanFord(source, target);
+    }
+
+    unsigned int cost = 0;
+    for (Vertex *v: vertexSet) {
+        for (Edge *e: v->getAdj()) {
+            cost += e->getCost() * e->getFlow();
+        }
+    }
+    result.second = cost;
+
+    return result;
```

Comment:
How can we relate this data? The Bellman-Ford gives out a single path, and you get the cost for that path. But then the flow you found is refering to all paths possible. Please say if you don't quite get my reasoning.
I think we could try to relate the cost and flow for each path. (The flow can be found by finding the bottleneck of that path)
Tell me what you think about this approach

Reply:
I am not quite getting what you're suggesting, but the idea of this algorithm is the following:
- Get regular maxFlow flows using EdmondsKarp
- Check residual and find negative cost cycles. Sending flow along a cycle in the residual graph always leads to another valid flow and keeps the total flow constant. If the sum of the costs for the edges in the cycle is negative, this will reduce the overall cost and improve the solution. 
- After no more negative cycles are found, only then sum up the cost of all the edges that have flow (cost is per unit of flow)

Reply:
After sleeping on it, I think getting the minCost and maxFlow between two stations is enough. Even though you can't quite compare them (if the minCost is really just the cost of the shortest path)